python/ml4ir/config/features.py

```diff
@@ -1,120 +1,237 @@
+
+        try:
+            self.features: List = features_dict.get(FeatureConfigKey.FEATURES)
+            self.all_features.extend(self.features)
+        except KeyError:
+            raise KeyError("'features' key not found in the feature_config specified")
+
+        # Features that can be used for training the model
+        self.ranking_features = list()
+        # Features that provide additional information about the query+records
+        self.metadata_features = list()
+        # Features that contain information at the query level common to all records
+        self.context_features = list()
+        # Features that contain information at the record level
+        self.sequence_features = list()
+        # Features to log at inference time
+        self.features_to_log = list()
+        for feature_info in self.all_features:
+            if feature_info.get("trainable", True):
+                self.ranking_features.append(feature_info)
+            else:
+                self.metadata_features.append(feature_info)
+
+            if feature_info.get("tfrecord_type") == TFRecordTypeKey.CONTEXT:
+                self.context_features.append(feature_info)
+            elif feature_info.get("tfrecord_type") == TFRecordTypeKey.SEQUENCE:
+                self.sequence_features.append(feature_info)
+            else:
+                raise TypeError(
+                    "tfrecord_type should be either context or sequence "
+                    "but is {} for {}".format(
+                        feature_info.get("tfrecord_type"), feature_info.get("name")
+                    )
+                )
+
+            if feature_info.get("log_at_inference", False):
+                self.features_to_log.append(feature_info)
+
+        if len(self.ranking_features) == 0:
             raise Exception("No trainable features specified in the feature config")
-        if not self.label:
-            raise Exception("No label field defined in the feature config")
 
-    def get_feature_names(self, trainable=None, include_label=False, dtype=None) -> List[str]:
+        if logger:
+            logger.info("Feature config loaded successfully")
+            logger.info(
+                "Trainable Features : \n{}".format("\n".join(self.get_ranking_features("name")))
+            )
+            logger.info("Click label : {}".format(self.get_label("name")))
+            logger.info(
+                "Metadata Features : \n{}".format("\n".join(self.get_metadata_features("name")))
+            )
+            logger.info(
+                "Context Features : \n{}".format("\n".join(self.get_ranking_features("name")))
+            )
+            logger.info(
+                "Sequence Features : \n{}".format("\n".join(self.get_ranking_features("name")))
+            )
+
+        self.mask = self.generate_mask()
+        self.all_features.append(self.get_mask())
+
+    def _process_dict(self, dict_, key: str = None):
```

Comment:
is "_process_dict" really the best name for this?  What is the type of this method, even?  It's the value of a the dict at a key, unless it's "node_name", in which case grab the "name", but if None is passed for the key, just return the dict?  

I'm confused.

Reply:
You either get the whole dictionary, or a particular value if you have specified the key. I'm using this as a wrapper for `ranking_features`, `metadata_features`, `context_features`, `sequence_features`..
This method is needed because I have to define `node_name` with the default value of `name`

Reply:
`_process_dict` is not a good name. I agree. But then again, I'm not sure what is. Do you have any suggestions?

Reply:
It is a fancy `.get()`

Reply:
`get_key_or_name` maybe?